src/middlewared/middlewared/plugins/replication.py

```diff
@@ -1,393 +1,440 @@
-            new['remote'],
-            repl_remote_dict
-        )
+        if verrors:
+            raise verrors
 
-        await self._service_change('ssh', 'reload')
+        periodic_snapshot_tasks = new["periodic_snapshot_tasks"]
+        await self.compress(new)
 
         await self.middleware.call(
-            'datastore.update',
+            "datastore.update",
             self._config.datastore,
             id,
             new,
             {'prefix': self._config.datastore_prefix}
         )
 
+        await self._set_periodic_snapshot_tasks(id, periodic_snapshot_tasks)
+
+        await self.middleware.call("service.restart", "cron")
+        await self.middleware.call("zettarepl.update_tasks")
+
         return await self._get_instance(id)
 
     @accepts(
-        Int('id')
+        Int("id")
     )
     async def do_delete(self, id):
-
-        replication = await self._get_instance(id)
-
-        try:
-            if replication['lastsnapshot']:
-                zfsname = replication['lastsnapshot'].split('@')[0]
-                await self.middleware.call('notifier.zfs_dataset_release_snapshots', zfsname, True)
-        except Exception:
-            pass
-
-        await self.middleware.call('replication.remove_from_state_file', id)
-
         response = await self.middleware.call(
-            'datastore.delete',
+            "datastore.delete",
             self._config.datastore,
             id
         )
 
-        await self._service_change('ssh', 'reload')
-
         return response
 
-    @private
-    def remove_from_state_file(self, id):
-        if os.path.exists(REPL_RESULTFILE):
-            with open(REPL_RESULTFILE, 'rb') as f:
-                data = f.read()
-            try:
-                results = pickle.loads(data)
-                results.pop(id, None)
-                with open(REPL_RESULTFILE, 'wb') as f:
-                    f.write(pickle.dumps(results))
-            except Exception as e:
-                self.logger.debug('Failed to remove replication from state file %s', e)
-
-        progressfile = '/tmp/.repl_progress_%d' % id
-        try:
-            os.unlink(progressfile)
-        except Exception:
-            pass
-
-    @accepts()
-    def public_key(self):
-        """
-        Get the public SSH replication key.
-        """
-        if (os.path.exists(REPLICATION_KEY) and os.path.isfile(REPLICATION_KEY)):
-            with open(REPLICATION_KEY, 'r') as f:
-                key = f.read()
+    async def _validate(self, data):
+        verrors = ValidationErrors()
+
+        # Direction
+
+        snapshot_tasks = []
+
+        if data["direction"] == "PUSH":
+            e, snapshot_tasks = await self._query_periodic_snapshot_tasks(data["periodic_snapshot_tasks"])
+            verrors.add_child("periodic_snapshot_tasks", e)
+
+            if data["naming_schema"]:
+                verrors.add("naming_schema", "This field has no sense for push replication")
```

Comment:
We should use `f"{schema_name}.attribute_name` in all verrors here, no?

Reply:
I am using this as:
```
        verrors = ValidationErrors()
        verrors.add_child("replication_update", await self._validate(new))
```
Don't need to pass schema name to validation method, code seems cleaner.

Reply:
Hm I am missing something, do you mean you will use that?

In that example you're using straight verrors.add which wont have the schema name.

Reply:
No. The code above is

`async def _validate(self, data):`

In `do_update` I use it as

```
        verrors = ValidationErrors()
        verrors.add_child("replication_update", await self._validate(new))
```

And `add_child` is

```
    def add_child(self, attribute, child):
        for e in child.errors:
            self.add(f"{attribute}.{e.attribute}", e.errmsg, e.errno)
```

So `_validate` returns `verrors` with names like `naming_schema` but `do_update` adds them to it's own `ValidationErrors` object as `replication_update.naming_schema`

Reply:
Ah, I get it, you're returning verrors and then adding it as a child later. Nevermind :D